### ISSUE TITLE
Each matches single Regex twice

### ISSUE THREAD

**Issue by scauligi** [pyparsing/pyparsing#183]
A Regex inside of an Each can incorrectly match twice.
This also applies to derivative expressions, such as `(Regex(...) | Regex(...))`

Minimum viable example:
```python
import pyparsing as pp

integer = pp.Regex(r'\d+')
each = pp.Each([integer])

# prints ['22', '33'] instead of ['22']
print(each.parseString('22 33'))

# this prints ['22'] as expected
print(pp.Each([pp.Word(pp.nums)]).parseString('22 33'))
```

The issue seems to be that `Regex` has `.mayReturnEmpty = True`.

In the `parseImpl` definition for `Each`, any expression that is not `Optional` but has `.mayReturnEmpty = True` gets assigned both to `self.optionals` and `self.required`;
and hence the expression is matched twice.

**Comment by ptmcg** [comment]
Thank you and nice catch (and diagnosis). At first glance, I think the solution will be to have `Regex` try to match `""` to see if it should set `mayReturnEmpty` to True or not. If it is possible for a particular `Regex` expression to match `""`, then it probably should remain True; otherwise, definitely set to False.

**Comment by ptmcg** [comment]
Ok! I have a solution with unit tests, so I'll be able to check into GitHub shortly. How urgent is this fix for you? Can you wait for 3.0.0?

**Comment by scauligi** [comment]
Thanks for the quick response! This isn't urgent, so I don't mind waiting.

**Comment by ptmcg** [comment]
Released in 3.0.0a2
